xcode/Subconscious/Shared/Services/NoosphereService.swift

```diff
@@ -0,0 +1,201 @@
+            )
+        }
+        self.noosphere = noosphere
+    }
+    
+    /// Create and configure a user and sphere
+    /// - Important: the identity should be persisted
+    /// - Important: the mnemonic should be displayed to user and then discarded.
+    /// - Returns: a SphereConfig containing the sphere identity and mnemonic
+    ///
+    /// What it does:
+    /// - Initializes a namespace
+    /// - Creates a key
+    /// - Creates a sphere for that key
+    public func createSphere(
+        ownerKeyName: String
+    ) throws -> SphereReceipt {
+        ns_key_create(noosphere, ownerKeyName)
+        
+        guard let sphereReceipt = ns_sphere_create(
+            noosphere,
+            ownerKeyName
+        ) else {
+            throw NoosphereError.foreignError(
+                "Failed to get pointer for sphere receipt"
+            )
+        }
+        defer {
+            ns_sphere_receipt_free(sphereReceipt)
+        }
+        
+        guard let sphereIdentityPointer = ns_sphere_receipt_identity(
+            sphereReceipt
+        ) else {
+            throw NoosphereError.foreignError(
+                "Failed to get pointer for identity"
+            )
+        }
+        defer {
+            ns_string_free(sphereIdentityPointer)
+        }
+        
+        guard let sphereMnemonicPointer = ns_sphere_receipt_mnemonic(
+            sphereReceipt
+        ) else {
+            throw NoosphereError.foreignError(
+                "Failed to get pointer for mnemonic"
+            )
+        }
+        defer {
+            ns_string_free(sphereMnemonicPointer)
+        }
+        
+        let sphereIdentity = String.init(cString: sphereIdentityPointer)
+        let sphereMnemonic = String.init(cString: sphereMnemonicPointer)
+        
+        return SphereReceipt(
+            identity: sphereIdentity,
+            mnemonic: sphereMnemonic
+        )
+    }
+    
+    deinit {
+        ns_free(noosphere)
+    }
+}
+
+public final class Sphere {
+    private let noosphere: Noosphere
+    let fs: OpaquePointer
+    let identity: String
+    
+    init(noosphere: Noosphere, identity: String) throws {
+        self.noosphere = noosphere
+        self.identity = identity
+        
+        guard let fs = ns_sphere_fs_open(
+            noosphere.noosphere,
+            identity
+        ) else {
+            throw NoosphereError.foreignError("Failed to get pointer for sphere file system")
+        }
+        self.fs = fs
+    }
+    
+    /// Read the value of a memo from a Sphere
+    /// - Returns: `SphereMemo`
+    func read(slashlink: String) throws -> SphereMemo {
+        guard let file = ns_sphere_fs_read(
+            noosphere.noosphere,
+            fs,
+            slashlink
+        ) else {
+            throw NoosphereError.memoDoesNotExist
+        }
+        defer {
+            ns_sphere_file_free(file)
+        }
+        
+        let contentTypeValues = ns_sphere_file_header_values_read(
```

Comment:
No change required, but do note that there may be more than one value in the headers. On the JS side we added a "get first header with this value" (which we also have in Rust). We can add it to the FFI too.

Reply:
Good to know. The Rust code says this returns an array of strings. However, on the Swift side we take the return value and convert it to a string. Should I expect problems with this? https://github.com/subconsciousnetwork/subconscious/pull/389/files/88332833ca3289deee3cb01591da2aae812fa924#diff-afa6d978b9708e5d6b6b9da40aad2dc1418053e91d037b458866ea22921bc2a9R158

Reply:
>  Should I expect problems with this?

Yes, probably! We should give you a way to get the first value.

Reply:
That would be great! Will land this for now and circle back when we have these new FFI functions.

Reply:
subconsciousnetwork/noosphere#199

Reply:
Created issue https://github.com/subconsciousnetwork/noosphere/issues/201